pkg/orchestrator/assetscanestimationwatcher/watcher.go

```diff
@@ -0,0 +1,261 @@
+			AssetScanEstimationID: assetScanEstimationID,
+			ScanEstimationID:      scanEstimationID,
+			AssetID:               assetID,
+		})
+	}
+
+	return events, nil
+}
+
+// nolint:cyclop
+func (w *Watcher) Reconcile(ctx context.Context, event AssetScanEstimationReconcileEvent) error {
+	logger := log.GetLoggerFromContextOrDiscard(ctx).WithFields(event.ToFields())
+	ctx = log.SetLoggerForContext(ctx, logger)
+
+	assetScanEstimation, err := w.backend.GetAssetScanEstimation(ctx, event.AssetScanEstimationID, models.GetAssetScanEstimationsAssetScanEstimationIDParams{
+		Expand: utils.PointerTo("asset"),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to get AssetScanEstimation with %s id: %w", event.AssetScanEstimationID, err)
+	}
+
+	state, ok := assetScanEstimation.GetState()
+	if !ok {
+		return fmt.Errorf("cannot determine state of AssetScanEstimation with %s id", event.AssetScanEstimationID)
+	}
+
+	logger.Tracef("Reconciling AssetScanEstimation state: %s", state)
+
+	switch state {
+	case models.AssetScanEstimationStateStatePending:
+		if err = w.reconcilePending(ctx, &assetScanEstimation); err != nil {
+			return err
+		}
+	case models.AssetScanEstimationStateStateAborted:
+		if err = w.reconcileAborted(ctx, &assetScanEstimation); err != nil {
+			return err
+		}
+	default:
+	}
+
+	return nil
+}
+
+// nolint:cyclop
+func (w *Watcher) reconcilePending(ctx context.Context, assetScanEstimation *models.AssetScanEstimation) error {
+	logger := log.GetLoggerFromContextOrDiscard(ctx)
+
+	logger.Debugf("reconciling pending asset scan estimations")
+
+	assetScanEstimationID, ok := assetScanEstimation.GetID()
+	if !ok {
+		return errors.New("invalid AssetScanEstimation: ID is nil")
+	}
+
+	if assetScanEstimation.Asset == nil || assetScanEstimation.Asset.AssetInfo == nil {
+		return errors.New("invalid AssetScanEstimation: Asset or AssetInfo is nil")
+	}
+
+	asset, err := assetScanEstimation.Asset.ToAsset()
+	if err != nil {
+		return fmt.Errorf("failed to convert assetRelationship to asset: %v", err)
+	}
+
+	stats := w.getLatestAssetScanStats(ctx, asset)
+
+	estimation, err := w.provider.Estimate(ctx, stats, asset, assetScanEstimation.AssetScanTemplate)
+
+	var fatalError provider.FatalError
+	var retryableError provider.RetryableError
+	switch {
+	case errors.As(err, &fatalError):
+		logger.Errorf("Failed to estimate asset scan: %v", err)
+		assetScanEstimation.State.State = utils.PointerTo(models.AssetScanEstimationStateStateFailed)
+		assetScanEstimation.State.StateMessage = utils.PointerTo(fatalError.Error())
+		assetScanEstimation.State.StateReason = utils.PointerTo(models.AssetScanEstimationStateStateReasonUnexpected)
+	case errors.As(err, &retryableError):
+		// nolint:wrapcheck
+		return common.NewRequeueAfterError(retryableError.RetryAfter(), retryableError.Error())
+	case err != nil:
+		logger.Errorf("Failed to estimate asset scan: %v", err)
+		assetScanEstimation.State.State = utils.PointerTo(models.AssetScanEstimationStateStateFailed)
```

Comment:
Do you think we should treat this same as retryableError?

Reply:
I saw in AssetScan watcher reconcileScheduled func that we are treating an error as fatal error and just marking it as done. Do we need to fix this in there as well?

Reply:
IMO yes it should be fixed there too, there is suppose to be three error cases:

* Fatal error - stop now, we know this isn't something a retry will fix

* Retryable/Requeue error - a known case were we need to re-run reconciliation, the raiser can specify how long to wait until its requeue. (IMO we should change the name of this, I don't like "Retryable" its confusing and doesn't fix the use cases its designed for such as waiting for a snapshot to be ready, perhaps RetryAfter or ReconcileAfter @chrisgacsal WDYT?)

* Normal error - an error occurred, it will be logged we'll retry after the next polling interval. We'll keep retrying until the timeout is hit.